---
title: "00.Fig.fibroblasts"
---

```{r knitr, include = FALSE}
DOCNAME <- "00.Fig.fibroblasts"
dir.create(here::here("output", DOCNAME), showWarnings = FALSE)

NOW <- Sys.time()
# Time chunks during knitting
knitr::knit_hooks$set(timeit = function(before) {
  if (before) {
    print(paste("Start:", Sys.time()))
    NOW <<- Sys.time()
  } else {
    print(paste("Stop:", Sys.time()))
    print(Sys.time() - NOW)
  }
})

knitr::opts_chunk$set(
  autodep        = TRUE,
  cache          = FALSE,
  cache.path     = paste0("cache/", DOCNAME, "/"),
  cache.lazy     = FALSE,
  cache.comments = FALSE,
  echo           = FALSE,
  error          = FALSE,
  fig.align      = "center",
  fig.width      = 10,
  fig.height     = 8,
  message        = FALSE,
  warning        = FALSE,
  timeit         = TRUE
)
```

## Setup

Load packages

```{r libaries, cache = FALSE, message=FALSE, warning=FALSE}
# Tidyverse
library(tidyverse)

# Plotting
library(scales)
library(EnvStats)
library(ggplotify)
library(ggcorrplot)
library(ggpubr)
library(ggrepel)
library(ggrastr)

# patch
library(patchwork)
library(cowplot)
my_theme <- theme_cowplot(
  font_size = 6, # Overall font size
  rel_small = 6 / 6, # axis tick labels
  rel_tiny = 6 / 6, # caption
  rel_large = 6 / 6, # title
  font_family = "Arial"
)
geom.text.size <- 6 / (14 / 5)

# heatmap
library(pheatmap)
library(ComplexHeatmap)
library(circlize)

# color
library(ggsci)

# Seurat
library(Seurat)

# monocle
library(monocle3)

# enrichment
library(clusterProfiler)
library(org.Hs.eg.db)
```

Load code

```{r source, cache = FALSE}
source(here::here("code/preprocess.R"))
source(here::here("code/glm.R"))
source(here::here("code/color.R"))
source(here::here("code/plot.R"))
```

## Load data

Load srat

```{r load-rds, cache=FALSE}
srat <- readRDS(here::here("output/04.rm_cells/seurat_fibroblasts.rds"))
Idents(srat) <- "level_3"
srat
```

Load metadata

```{r load-meta, cache=FALSE, message=FALSE, warning=FALSE}
metadata <- read_csv(here::here("output/04.rm_cells/seurat_fibroblasts.cellmeta.csv")) %>%
  mutate(cellType3 = level_3)
sam_info <- read_csv(here::here("data/sample_info.3.csv"))
```

Load c2c

```{r load-c2c, cache=FALSE}
df_c2c <- readRDS(here::here("data", "c2c.fib.rds"))
head(df_c2c)
```

## UMAP {.tabset}

### Level 3

```{r fig.width=3, fig.height=2.5, cache=FALSE}
p.umap.cellType <- DimPlot(srat,
  reduction = "umap", group.by = "level_3",
  raster = TRUE
) +
  scale_color_manual(values = fib_level_3_color) +
  labs(title = NULL) +
  my_theme
p.umap.cellType

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.umap.cellType.pdf"),
  height = 2.5, width = 3, device = cairo_pdf
)
```

With label

```{r fig.width=2, fig.height=2, cache=FALSE}
p.umap.celltype.label <- DimPlot(srat,
  reduction = "umap", group.by = "level_3",
  label = TRUE, label.size = geom.text.size,
  raster = TRUE
) +
  scale_color_manual(values = fib_level_3_color) +
  labs(title = NULL) +
  my_theme +
  theme(legend.position = "none")
p.umap.celltype.label

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.umap.cellType.labelled.pdf"),
  height = 2, width = 2, device = cairo_pdf
)
```

## Classification

### Subtypes

```{r fig.width=2.8, fig.height=3}
markers <- list(
  NMF = filter_features(srat, c("SLPI", "PI16", "TNXB", "IGFBP6", "ADH1B")),
  iCAF = filter_features(srat, c(
    "IL6", "CXCL12",
    "HAS1", "CXCL1", "CCL2"
  )),
  myCAF = filter_features(srat, c(
    "ACTA2", "COL1A2", "TAGLN", "MMP11", "POSTN"
  )),
  pericytes = filter_features(srat, c("RGS5", "MCAM", "PDGFRB", "COL4A1")),
  apCAF = filter_features(srat, c("CD74", "HLA-DRA", "HLA-DPA1")),
  VSMC = filter_features(srat, c("MYH11", "DES", "ACTG2"))
)

mat <- AverageExpression(srat,
  assays = "RNA", features = unlist(markers),
  slot = "data", verbose = F
)$RNA
mat <- log1p(as.matrix(mat))
# scale
mat_scale <- pheatmap:::scale_rows(mat)

# color
x <- ceiling(max(abs(mat_scale)))
col_fun <- circlize::colorRamp2(c(-x, 0, x), c("#476fa9", "#ffffff", "#ca3226"))
legend_name <- "Exp."

ht <- ComplexHeatmap::Heatmap(mat_scale,
  name = legend_name,
  col = col_fun,
  cluster_columns = FALSE, cluster_rows = FALSE,
  show_row_names = TRUE, show_column_names = TRUE,
  row_names_side = "right",
  row_names_gp = gpar(fontsize = 6, fontface = "italic"),
  column_names_gp = gpar(fontsize = 6),
  column_names_rot = 60,
  row_split = factor(
    rep(names(markers), sapply(markers, length)),
    levels = names(markers)
  ),
  row_title_gp = gpar(fontsize = 6),
  column_split = factor(c(
    rep(c("NMF"), 3),
    rep(c("iCAF"), 3),
    rep(c("myCAF"), 3),
    rep(c("pericyte"), 3),
    "apCAF",
    "VSMC"
  ),
  levels = levels(srat$level_2)
  ),
  column_title = NULL,
  left_annotation = rowAnnotation(
    empty = anno_empty(
      border = FALSE,
      width = unit(1, "mm")
    ),
    foo = anno_block(
      gp = gpar(
        fill = "gray",
        col = "gray"
      ),
      width = unit(0.5, "mm")
    )
  ),
  heatmap_legend_param = list(
    title_position = "topcenter",
    direction = "vertical",
    labels_gp = gpar(fontsize = 6),
    title_gp = gpar(fontsize = 8),
    grid_height = unit(1.5, "mm"),
    grid_width = unit(1.5, "mm")
  )
)

# draw(ht)
p.markers.canonical <- plot_grid(grid.grabExpr(draw(ht)))
p.markers.canonical

# save
saveRDS(p.markers.canonical, file = here::here("output", DOCNAME, "Fig.markers.canonical.rds"))
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.markers.canonical.pdf"),
  height = 3, width = 2.8, device = cairo_pdf
)
```

Small

```{r fig.width=2, fig.height=2}
ht <- ComplexHeatmap::Heatmap(mat_scale,
  name = legend_name,
  col = col_fun,
  cluster_columns = FALSE, cluster_rows = FALSE,
  show_row_names = TRUE, show_column_names = TRUE,
  row_names_side = "right",
  row_names_gp = gpar(fontsize = 4, fontface = "italic"),
  column_names_gp = gpar(fontsize = 4),
  column_names_rot = 60,
  row_split = factor(
    rep(names(markers), sapply(markers, length)),
    levels = names(markers)
  ),
  row_title_gp = gpar(fontsize = 4),
  column_split = factor(c(
    rep(c("NMF"), 3),
    rep(c("iCAF"), 3),
    rep(c("myCAF"), 3),
    rep(c("pericyte"), 3),
    "apCAF",
    "VSMC"
  ),
  levels = levels(srat$level_2)
  ),
  column_title = NULL,
  left_annotation = rowAnnotation(
    empty = anno_empty(
      border = FALSE,
      width = unit(1, "mm")
    ),
    foo = anno_block(
      gp = gpar(
        fill = "gray",
        col = "gray"
      ),
      width = unit(0.5, "mm")
    )
  ),
  heatmap_legend_param = list(
    title_position = "topcenter",
    direction = "vertical",
    labels_gp = gpar(fontsize = 4),
    title_gp = gpar(fontsize = 6),
    grid_height = unit(1.5, "mm"),
    grid_width = unit(1.5, "mm")
  )
)

# draw(ht)
p.markers.canonical <- plot_grid(grid.grabExpr(draw(ht)))
p.markers.canonical

# save
saveRDS(p.markers.canonical, file = here::here("output", DOCNAME, "Fig.markers.canonical.small.rds"))
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.markers.canonical.small.pdf"),
  height = 2, width = 2, device = cairo_pdf
)
```

## Origin2 enrichment

```{r fig.width=1.8, fig.height=2.5, warning=FALSE}
cluster_cellNum <- metadata %>%
  group_by(cellType3) %>%
  summarise(cluster.cell_num = n()) %>%
  mutate(cluster.cell_num = as.numeric(cluster.cell_num))

origin2_CellNum <- sam_info %>%
  group_by(Origin2_n) %>%
  summarise(origin.cell_num = sum(cell.filtered)) %>%
  mutate(origin.cell_num = as.numeric(origin.cell_num)) %>%
  mutate(total = sum(origin.cell_num)) %>%
  mutate(total = as.numeric(total))

cluster_byOrigin2 <- metadata %>%
  group_by(cellType3, Origin2_n) %>%
  summarise(o = n()) %>%
  mutate(o = as.numeric(o))

# add origins not existing
tmp <- setdiff(unique(metadata$Origin2_n), cluster_byOrigin2$Origin2_n)
if (length(tmp) > 0) {
  for (t in tmp) {
    x <- tibble(cellType3 = unique(cluster_cellNum$cellType3), Origin2_n = t, o = 0)
    cluster_byOrigin2 <- bind_rows(cluster_byOrigin2, x)
  }
}

r_oe <- cluster_byOrigin2 %>%
  left_join(origin2_CellNum, by = "Origin2_n") %>%
  left_join(cluster_cellNum, by = "cellType3") %>%
  mutate(e = (origin.cell_num * cluster.cell_num) / total) %>%
  mutate(r = o / e) %>%
  dplyr::select(cellType3, Origin2_n, r) %>%
  spread(Origin2_n, r, fill = 0) %>%
  column_to_rownames("cellType3")

# order
tmp.df <- r_oe %>%
  rownames_to_column("cellType3") %>%
  arrange(nLN, pLN, Normal, Adjacent, Tumor)

# data for plot
r_oe[r_oe > 10] <- 10
d4p <- r_oe %>%
  rownames_to_column("cellType3") %>%
  pivot_longer(cols = !cellType3, names_to = "Origin2_n") %>%
  mutate(
    Origin2_n = factor(Origin2_n, levels = c(
      "prePBMC", "postPBMC",
      "nLN", "pLN",
      "Normal", "Adjacent",
      "Tumor"
    )),
    cellType3 = factor(cellType3, levels = tmp.df$cellType3)
  )

# % lineage prop
df.lineage_prop <- metadata %>%
  group_by(cellType3, Origin2_n) %>%
  summarise(cellType3.cell_num = n()) %>%
  mutate(cellType3.cell_num = as.numeric(cellType3.cell_num)) %>%
  ungroup() %>%
  mutate(tot = nrow(metadata)) %>%
  mutate(lineage.prop = cellType3.cell_num * 100 / tot)

# plot
p.roe.fib <- d4p %>%
  filter(!str_detect(Origin2_n, "PBMC")) %>%
  ggplot(aes(x = Origin2_n, y = cellType3, fill = value)) +
  geom_tile() +
  geom_point(
    data = df.lineage_prop,
    aes(x = Origin2_n, y = cellType3, size = lineage.prop),
    shape = 1,
    inherit.aes = FALSE
  ) +
  scale_x_discrete(position = "top") +
  scale_y_discrete(position = "right") +
  scale_fill_gradient2(
    midpoint = 1,
    low = "blue", mid = "white", high = "red",
    name = expression(R[O / E])
  ) +
  scale_size(range = c(1, 3)) +
  labs(size = "% lineage") +
  theme_cowplot(font_size = 6) +
  theme(
    axis.text.x = element_text(angle = 60, hjust = 0),
    axis.title = element_blank(),
    axis.line = element_blank(),
    axis.ticks = element_blank(),
    legend.position = "right",
    text = element_text(size = 6)
  )
p.roe.fib

# save
saveRDS(p.roe.fib, file = here::here("output", DOCNAME, "Fig.Roe.byOrigin2_n.rds"))
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.Roe.byOrigin2_n.pdf"),
  height = 2.5, width = 1.8, device = cairo_pdf
)
```

## monocle3

Load cds

```{r load-cds}
cds <- readRDS(here::here("output/04.rm_cells/monocle3_fibroblasts.rds"))
cds.par_1 <- readRDS(here::here("output/monocle3.fib.par_1/monocle3_fibroblasts.par_1.rds"))
cds.par_2 <- readRDS(here::here("output/monocle3.fib.par_2/monocle3_fibroblasts.par_2.rds"))
```

trajectory

```{r fig.width=3, fig.height=3}
p.monocle3_trajectory <- plot_cells(cds,
  label_groups_by_cluster = FALSE,
  color_cells_by = "partition",
  label_leaves = FALSE,
  label_branch_points = TRUE
) +
  my_theme +
  theme(legend.position = c(.1, .85))
p.monocle3_trajectory

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_trajectory.pdf"),
  height = 3, width = 3, device = cairo_pdf
)
```

### NMF + iCAF

trajectory

```{r fig.width=3, fig.height=3}
p.monocle3_trajectory.par1 <- plot_cells(cds.par_1,
  label_groups_by_cluster = FALSE,
  color_cells_by = "level_3",
  label_leaves = FALSE,
  cell_size = .5,
  group_label_size = 3,
  label_branch_points = FALSE
) +
  # labs(title = "Monocle3") +
  scale_color_manual(values = fib_level_3_color) +
  my_theme +
  theme(legend.position = "none")
p.monocle3_trajectory.par1

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_trajectory.NMF_to_iCAF.pdf"),
  height = 3, width = 3, device = cairo_pdf
)
```

Pseudotime

```{r fig.width=3.2, fig.height=3}
p <- plot_cells(cds.par_1, color_cells_by = "pseudotime",
           label_cell_groups = FALSE, label_leaves = FALSE, 
           label_branch_points = TRUE) +
  my_theme
p

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_pseudotime.NMF_to_iCAF.pdf"),
  height = 3, width = 3.2, device = cairo_pdf
)
```

tissue

```{r fig.width=3.5, fig.height=3}
p.monocle3_tissue.par1 <- plot_cells(cds.par_1,
  label_groups_by_cluster = FALSE,
  color_cells_by = "Origin2_n",
  label_leaves = FALSE,
  cell_size = .5,
  label_cell_groups = FALSE,
  #group_label_size = 3,
  label_branch_points = FALSE,
  show_trajectory_graph = FALSE
) +
  scale_color_manual(values = origin2_color_maps) +
  my_theme
p.monocle3_tissue.par1

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.umap_origin2.NMF_to_iCAF.pdf"),
  height = 3, width = 3.5, device = cairo_pdf
)
```


SOX4

```{r fig.width=2, fig.height=1.8}
srat.par_1 <- readRDS(here::here("output/monocle3.fib.par_1/seurat_fibroblasts.par_1.rds"))
FeaturePlot(srat.par_1, features = "SOX4", pt.size = .1, order = T) +
  labs(title = NULL) +
  my_theme

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_genes.NMF-C1_to_iCAF-C1.SOX4.pdf"),
  width = 2, height = 1.8
)
```


NMF-C1_C2_C3_iCAF-C1 genes (no SOX4)

```{r fig.width=6, fig.height=6}
cds_sub = readRDS(here::here("output", "monocle3.fib.par_1.NMF-C1_C2_C3_iCAF-C1", "cds_sub.rds"))
ciliated_cds_pr_test_res = readRDS(here::here("output", "monocle3.fib.par_1.NMF-C1_C2_C3_iCAF-C1", "ciliated_cds_pr_test_res.rds"))

pr_graph_test_res = subset(ciliated_cds_pr_test_res, q_value < 0.01 & morans_I > 0.2)
pr_deg_ids <- row.names(pr_graph_test_res)

pt.matrix <- exprs(cds_sub)[match(pr_deg_ids, rownames(rowData(cds_sub))), order(pseudotime(cds_sub))]
#Can also use "normalized_counts" instead of "exprs" to use various normalization methods, for example:
#normalized_counts(cds, norm_method = "log")

pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x,df=3)$y}))
pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(pt.matrix) <- pr_deg_ids
x = pseudotime(cds_sub)
cells = names(x)[order(x)]
colnames(pt.matrix) = cells
pt.matrix = na.omit(pt.matrix)

# order
a = order(apply(pt.matrix, 1, FUN = function(i){
  which(i == max(i))
}))
pt.matrix = pt.matrix[a, ]

anno = as.data.frame(colData(cds_sub)[colnames(pt.matrix), c("Origin2_n", "level_3")])
anno = droplevels(anno)

gene_selected = c("JUN", "NGFR", "TGFBI", "IGFBP7", "APOD", "C7", "A2M", "SPARCL1", "CXCL14", "RGS5", "CTSL", "SOD2", "TNFRSF12A", "DDX21", "FGF7", "GEM", "IGF1", "IGFBP4", "BGN", "UGDH", "THBS1", "CXCL12", "SOD3", "SLPI", "DCN", "CFD", "CD99", "IGFBP6", "FOS", "JUNB", "TIMP3", "COL1A2", "GPX3", "GLUL", "ZFP36", "UAP1")

# marker
marker_2_lab = intersect(gene_selected, pr_deg_ids)
# location
marker_2_lab = sapply(marker_2_lab, function(x){which(rownames(pt.matrix) == x)}, simplify = T)
ha = rowAnnotation(foo = anno_mark(at = marker_2_lab,
                                   labels = names(marker_2_lab),
                                   labels_gp = gpar(fontsize = 6)))

col_ha = HeatmapAnnotation(
  df = anno,
  annotation_name_gp = gpar(fontsize = 6),
  col = list(
      Origin4 = origin2_color_maps[levels(anno$Origin2_n)],
      level_3 = fib_level_3_color[levels(anno$level_3)]
    )
)

hthc <- Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2, length=11),
                                            rev(brewer.pal(11, "Spectral"))),
  show_row_names               = FALSE,
  show_column_names            = FALSE,
  right_annotation = ha,
  top_annotation = col_ha,
  row_names_gp                 = gpar(fontsize = 8),
  row_title_rot                = 0,
  cluster_rows                 = FALSE,
  clustering_method_rows       = "ward.D2",
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE)
#draw(hthc)
p <- plot_grid(grid.grabExpr(draw(hthc)))
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_gene.NMF-C1_to_iCAF-C1.pdf"),
  height = 6, width = 6, device = cairo_pdf
)
```

NMF-C1_C2_iCAF-C3 genes (SOX4 in)

```{r fig.width=6, fig.height=6}
cds_sub = readRDS(here::here("output", "monocle3.fib.par_1.NMF-C1_C2_iCAF-C3", "cds_sub.rds"))
ciliated_cds_pr_test_res = readRDS(here::here("output", "monocle3.fib.par_1.NMF-C1_C2_iCAF-C3", "ciliated_cds_pr_test_res.rds"))
pr_graph_test_res = subset(ciliated_cds_pr_test_res, q_value < 0.01 & morans_I > 0.2)
pr_deg_ids <- row.names(pr_graph_test_res)

pt.matrix <- exprs(cds_sub)[match(pr_deg_ids, rownames(rowData(cds_sub))), order(pseudotime(cds_sub))]
#Can also use "normalized_counts" instead of "exprs" to use various normalization methods, for example:
#normalized_counts(cds, norm_method = "log")

pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x,df=3)$y}))
pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(pt.matrix) <- pr_deg_ids

x = pseudotime(cds_sub)
cells = names(x)[order(x)]
colnames(pt.matrix) = cells

pt.matrix = na.omit(pt.matrix)

# order
a = order(apply(pt.matrix, 1, FUN = function(i){
  which(i == max(i))
}))
pt.matrix = pt.matrix[a, ]

anno = as.data.frame(colData(cds_sub)[colnames(pt.matrix), c("Origin2_n", "level_3")])
anno = droplevels(anno)

gene_selected = c("UAP1", "CFD", "TIMP3", "SOD3", "IGFBP6", "CST3", "FOS", "GPX3", "ZFP36", "JUNB", "JUN", "FBN1", "DCN", "CD55", "MDK", "ISG15", "IGF1", "THBS2", "ADAM12", "BGN", "POSTN", "LOX", "RGS5", "IGFBP7", "SPARC", "CXCL2", "NR4A2", "HES1", "RUNX1", "IFI27", "IFI6", "CXCL14", "VCAN", "PDPN", "SOD2", "THBS1", "C7", "A2M", "SOX4")

# marker
marker_2_lab = intersect(gene_selected, pr_deg_ids)
# location
marker_2_lab = sapply(marker_2_lab, function(x){which(rownames(pt.matrix) == x)}, simplify = T)
ha = rowAnnotation(foo = anno_mark(at = marker_2_lab,
                                   labels = names(marker_2_lab),
                                   labels_gp = gpar(fontsize = 6)))

col_ha = HeatmapAnnotation(
  df = anno,
  annotation_name_gp = gpar(fontsize = 6),
  col = list(
      Origin4 = origin2_color_maps[levels(anno$Origin4)],
      level_3 = fib_level_3_color[levels(anno$level_3)]
    )
)

hthc <- Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2, length=11),
                                            rev(brewer.pal(11, "Spectral"))),
  show_row_names               = FALSE,
  show_column_names            = FALSE,
  right_annotation = ha,
  top_annotation = col_ha,
  row_names_gp                 = gpar(fontsize = 8),
  row_title_rot                = 0,
  cluster_rows                 = FALSE,
  clustering_method_rows       = "ward.D2",
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE)
#draw(hthc)
p <- plot_grid(grid.grabExpr(draw(hthc)))
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_gene.NMF-C1_to_iCAF-C3.pdf"),
  height = 6, width = 6, device = cairo_pdf
)
```

### pericyte + myCAF

trajectory

```{r fig.width=3, fig.height=3}
p.monocle3_trajectory.par2 <- plot_cells(cds.par_2,
  label_groups_by_cluster = FALSE,
  color_cells_by = "level_3",
  label_leaves = FALSE,
  cell_size = .5,
  group_label_size = 3,
  label_branch_points = FALSE
) +
  # labs(title = "Monocle3") +
  scale_color_manual(values = fib_level_3_color) +
  my_theme +
  theme(legend.position = "none")
p.monocle3_trajectory.par2

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_trajectory.pericyte_to_myCAF.pdf"),
  height = 3, width = 3, device = cairo_pdf
)
```

Pseudotime

```{r fig.width=3.1, fig.height=3}
p <- plot_cells(cds.par_2, color_cells_by = "pseudotime",
           label_cell_groups = FALSE, label_leaves = FALSE, 
           label_branch_points = TRUE) +
  my_theme
p

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_pseudotime.pericyte_to_myCAF.pdf"),
  height = 3, width = 3.1, device = cairo_pdf
)
```

tissue

```{r fig.width=3.5, fig.height=3}
p.monocle3_tissue.par2 <- plot_cells(cds.par_2,
  label_groups_by_cluster = FALSE,
  color_cells_by = "Origin2_n",
  label_leaves = FALSE,
  cell_size = .5,
  label_cell_groups = FALSE,
  #group_label_size = 3,
  label_branch_points = FALSE,
  show_trajectory_graph = FALSE
) +
  scale_color_manual(values = origin2_color_maps) +
  my_theme
p.monocle3_tissue.par2

# save
ggsave2(
  filename = here::here("output", DOCNAME, "Fig.umap_origin2.pericyte_2_myCAF.pdf"),
  height = 3, width = 3.5, device = cairo_pdf
)
```

SOX4

```{r fig.width=2, fig.height=1.8}
srat.par_2 <- readRDS(here::here("output/monocle3.fib.par_2/seurat_fibroblasts.par_2.rds"))
FeaturePlot(srat.par_2, features = "SOX4", pt.size = .1) +
  labs(title = NULL) +
  my_theme

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_genes.pericyte-C1_to_myCAF-C2.SOX4.pdf"),
  width = 2, height = 1.8
)
```

Genes (no SOX4)

```{r fig.width=6, fig.height=6}
cds_sub = readRDS(here::here("output", "monocle3.fib.par_2.pericyte-C1_C3_myCAF-C2", "cds_sub.rds"))
ciliated_cds_pr_test_res = readRDS(here::here("output", "monocle3.fib.par_2.pericyte-C1_C3_myCAF-C2", "ciliated_cds_pr_test_res.rds"))

pr_graph_test_res = subset(ciliated_cds_pr_test_res, q_value < 0.01 & morans_I > 0.2)
pr_deg_ids <- row.names(pr_graph_test_res)
pt.matrix <- exprs(cds_sub)[match(pr_deg_ids, rownames(rowData(cds_sub))), order(pseudotime(cds_sub))]
#Can also use "normalized_counts" instead of "exprs" to use various normalization methods, for example:
#normalized_counts(cds, norm_method = "log")

pt.matrix <- t(apply(pt.matrix,1,function(x){smooth.spline(x,df=3)$y}))
pt.matrix <- t(apply(pt.matrix,1,function(x){(x-mean(x))/sd(x)}))
rownames(pt.matrix) <- pr_deg_ids

x = pseudotime(cds_sub)
cells = names(x)[order(x)]
colnames(pt.matrix) = cells

pt.matrix = na.omit(pt.matrix)

# order
a = order(apply(pt.matrix, 1, FUN = function(i){
  which(i == max(i))
}))
pt.matrix = pt.matrix[a, ]
pt.matrix[1:3, 1:3]

anno = as.data.frame(colData(cds_sub)[colnames(pt.matrix), c("Origin2_n", "level_3")])
anno = droplevels(anno)

gene_selected = c("SPARC", "PLOD2", "TPM2", "SULF1", "THBS2", "VCAN", "CD99", "FAP", "GJA1", "FN1", "BGN", "IFI6", "LY6E", "SOD2", "CTSB", "CTSK", "PDPN", "ANTXR1", "PDGFRL", "ISG15", "YBX1", "CD36", "IGFBP7", "CXCL2", "FABP4", "ADAMTS4", "CAV1", "RGS5", "GJA4", "SPARCL1", "CDKN1A", "IL6", "MTHFD2")

# marker
marker_2_lab = intersect(gene_selected, pr_deg_ids)
# location
marker_2_lab = sapply(marker_2_lab, function(x){which(rownames(pt.matrix) == x)}, simplify = T)
ha = rowAnnotation(foo = anno_mark(at = marker_2_lab,
                                   labels = names(marker_2_lab),
                                   labels_gp = gpar(fontsize = 6)))

col_ha = HeatmapAnnotation(
  df = anno,
  annotation_name_gp = gpar(fontsize = 6),
  col = list(
      Origin4 = origin2_color_maps[levels(anno$Origin2_n)],
      level_3 = fib_level_3_color[levels(anno$level_3)]
    )
)

hthc <- Heatmap(
  pt.matrix,
  name                         = "z-score",
  col                          = colorRamp2(seq(from=-2,to=2, length=11),
                                            rev(brewer.pal(11, "Spectral"))),
  show_row_names               = FALSE,
  show_column_names            = FALSE,
  right_annotation = ha,
  top_annotation = col_ha,
  row_names_gp                 = gpar(fontsize = 8),
  row_title_rot                = 0,
  cluster_rows                 = FALSE,
  clustering_method_rows       = "ward.D2",
  cluster_row_slices           = FALSE,
  cluster_columns              = FALSE)
#draw(hthc)
p <- plot_grid(grid.grabExpr(draw(hthc)))
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.monocle3_gene.pericyte-C1_to_myCAF-C2.pdf"),
  height = 6, width = 6, device = cairo_pdf
)
```

## Selected genes

### Collagen

```{r collagen-exp, fig.width=2.8, fig.height=4}
markers <- list(
  Fibrillar = c("COL1A1", "COL1A2", "COL2A1", "COL3A1", "COL5A1", "COL5A2", "COL5A3", "COL11A1", "COL11A2"),
  Non_fibrillar = c("COL16A1", "COL8A2", "COL9A2", "COL24A1", "COL4A4", "COL4A3", "COL6A3", "COL7A1", "COL8A1", "COL6A5", "COL6A6", "COL25A1", "COL23A1", "COL21A1", "COL19A1", "COL9A1", "COL12A1", "COL10A1", "COL28A1", "COL26A1", "COL4A6", "COL4A5", "COL14A1", "COL22A1", "COL15A1", "COL27A1", "COL13A1", "COL17A1", "COL4A1", "COL4A2", "COL9A3", "COL18A1", "COL6A1", "COL6A2")
)

ht <- customDoHeatmap(srat, markers, z_score = T, idents = "level_3", cluster_columns = FALSE)
p <- plot_grid(grid.grabExpr(draw(ht, heatmap_legend_side = "right")))
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.exp_selected_genes.collagen.pdf"),
  height = 4, width = 2.8, device = cairo_pdf
)
```

### collagen-processing enzymes

- Paper: From: A. Sebastian, N. R. Hum, K. A. Martin, S. F. Gilmore, I. Peran, S. W. Byers, E. K. Wheeler, M. A. Coleman, G. G. Loots, Single-Cell Transcriptomic Analysis of Tumor-Derived Fibroblasts and Normal Tissue-Resident Fibroblasts Reveals Fibroblast Heterogeneity in Breast Cancer. Cancers. 12, 1307 (2020).

```{r Sebastian-collagen-processing-enzymes, fig.width=3, fig.height=2.5, cache=FALSE}
markers <- c(
  "ADAMTS2", "CTSK", "LOX", "LOXL1", "LOXL2", "LOXL3", "MMP2",
  "MMP3", "P3H1", "P3H3", "P3H4", "P4HA1", "P4HA2", "P4HA3", "P4HB", "PLOD1", "PLOD2", "PLOD3"
)

ht <- customDoHeatmap(srat, markers, z_score = T, idents = "level_3", cluster_columns = FALSE)
p <- plot_grid(grid.grabExpr(draw(ht, heatmap_legend_side = "right")))
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.exp_selected_genes.collagen-processing-enzymes.pdf"),
  height = 2.5, width = 3, device = cairo_pdf
)
```

### CAF – myeloid cell interactions

- Paper: Barrett, R., and Puré, E. (2020). Cancer-associated fibroblasts: key determinants of tumor immunity and immunotherapy. Current Opinion in Immunology 64, 80–87. (figure 1)

- CAFs secrete many signaling molecules known to influence both recruitment and activation state of myeloid cells including: CXCL1, CXCL2, CXCL5, CXCL6/GCP-2, CXCL8, CXCL9, CXCL10, CXCL12/SDF1, CCL2/MCP-1, CCL3, CCL5/Rantes, CCL7, CCL20, CCL26, IL-1b, IL-6, IL-10, VEGF, TGF-b, indoleamine-2,3-dioxygenase (IDO), prostaglandin (PG) E2 (PGE2), tumor necrosis factor (TNF) or nitric oxide (NO)

```{r Barrett-1, fig.height=3.5, fig.width=3, cache=FALSE}
markers <- c("CXCL1", "CXCL2", "CXCL5", "CXCL6", "CXCL8", "CXCL9", "CXCL10", "CXCL12", "CCL2", "CCL3", "CCL5", "CCL7", "CCL20", "CCL26", "IL1B", "IL6", "IL10", "VEGFA", "TGFB1", "IDO1", "PTGS2", "TNF", "NOS1", "NOS2", "NOS3")

ht <- customDoHeatmap(srat, markers, z_score = T, idents = "level_3", cluster_columns = FALSE)
p <- plot_grid(grid.grabExpr(draw(ht, heatmap_legend_side = "right")))
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.exp_selected_genes.CAF-myeloid-interaction.pdf"),
  height = 3.5, width = 3, device = cairo_pdf
)
```

### LRRC15+ myofibroblasts

LRRC15+ myofibroblasts dictate the stromal setpoint to suppress tumour immunity

发现TGFβ受体信号通路对于LRRC15+成纤维细胞细胞群促进癌症进展非常关键，通过减少LRRC15+成纤维细胞细胞群的治疗方法的开发可能会提高患者的生存以及对免疫治疗的反应。

```{r fig.height=3, fig.width=3}
p <- StackedVlnPlot(srat, features =  c("COL10A1", "COL11A1", "MMP11", "LRRC15"),
               color.use = fib_level_3_color, angle.x = 60)
p

ggsave2(
  filename = here::here("output", DOCNAME, "Fig.exp_selected_genes.LRRC15+myofibroblasts.pdf"),
  height = 3, width = 3, device = cairo_pdf
)
```

## cmp. subtypes

### pericyte-C3 vs pericyte-C1C2

```{r fig.width=2.5, fig.height=2}
deg <- read_csv(here::here('output', "06.fib.cmp_pericyte_C3_C1C2", 'deg.pericyte-C3_vs_pericyte-C1C2.csv'))
compareList = list(
  C1C2.up = deg %>% filter(p_val_adj < 0.01, avg_logFC < -0.25) %>% pull(gene) %>%
    bitr(fromType="SYMBOL", toType=c("ENTREZID"), OrgDb="org.Hs.eg.db") %>% pull(ENTREZID),
  C3.up = deg %>%filter(p_val_adj < 0.01, avg_logFC > 0.25) %>% pull(gene) %>%
    bitr(fromType="SYMBOL", toType=c("ENTREZID"), OrgDb="org.Hs.eg.db") %>% pull(ENTREZID)
)

cg.BP = compareCluster(compareList, fun = "enrichGO", OrgDb = org.Hs.eg.db, ont = "BP",
                       pAdjustMethod = "BH", pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = T)

d4p <- cg.BP
d4p@compareClusterResult <- d4p@compareClusterResult %>% 
  filter(Description %in% c("extracellular matrix organization",
                            "extracellular structure organization",
                            "collagen metabolic process"))


dotplot(d4p) + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 40)) + 
  scale_size(range = c(0, 5)) +
  my_theme +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

ggsave2(
  filename = here::here("output", DOCNAME, "deg.cmp_pericyte-C3_vs_pericyte-C1C2.GO-BP.pdf"),
  width = 2.5, height = 2, device = cairo_pdf
)
```

### iCAF-C3_vs_iCAF-C1C2

```{r fig.width=2.5, fig.height=2}
deg <- read_csv(here::here('output', "06.fib.cmp_iCAF_C3_C1C2", 'deg.iCAF-C3_vs_iCAF-C1C2.csv'))

compareList = list(
  C1C2.up = deg %>% filter(p_val_adj < 0.01, avg_logFC < -0.25) %>% pull(gene) %>%
    bitr(fromType="SYMBOL", toType=c("ENTREZID"), OrgDb="org.Hs.eg.db") %>% pull(ENTREZID),
  C3.up = deg %>%filter(p_val_adj < 0.01, avg_logFC > 0.25) %>% pull(gene) %>%
    bitr(fromType="SYMBOL", toType=c("ENTREZID"), OrgDb="org.Hs.eg.db") %>% pull(ENTREZID)
)
cg.BP = compareCluster(compareList, fun = "enrichGO", OrgDb = org.Hs.eg.db, ont = "BP",
                       pAdjustMethod = "BH", pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = T)

d4p <- cg.BP
d4p@compareClusterResult <- d4p@compareClusterResult %>% 
  filter(Description %in% c("extracellular matrix organization",
                            "extracellular structure organization"))
dotplot(d4p) + 
  scale_y_discrete(labels = function(x) str_wrap(x, width = 40)) + 
  my_theme +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

ggsave2(
  filename = here::here("output", DOCNAME, "deg.cmp_iCAF-C3_vs_iCAF-C1C2.GO-BP.pdf"),
  width = 2.5, height = 2, device = cairo_pdf
)
```




## Session info
