---
title: "monocle2.epithelia"
---

```{r knitr, include = FALSE}
DOCNAME = "monocle2.epithelia"
#dir.create(here::here("output", DOCNAME), showWarnings = FALSE)

NOW <- Sys.time()
# Time chunks during knitting
knitr::knit_hooks$set(timeit = function(before) {
    if (before) {
        print(paste("Start:", Sys.time()))
        NOW <<- Sys.time()
    } else {
        print(paste("Stop:", Sys.time()))
        print(Sys.time() - NOW)
    }
})

knitr::opts_chunk$set(
    autodep        = TRUE,
    cache          = FALSE,
    cache.path     = paste0("cache/", DOCNAME, "/"),
    cache.lazy     = FALSE, 
    cache.comments = FALSE,
    echo           = TRUE,
    error          = FALSE,
    fig.align      = "center",
    fig.width      = 10,
    fig.height     = 8,
    message        = FALSE,
    warning        = FALSE,
    timeit         = TRUE
)
```

## Setup

Load packages

```{r libaries, cache = FALSE}
# Tidyverse
library(tidyverse)
# Plotting
library(ggplotify)
library(ggcorrplot)
library(ggpubr)
library(ggrepel)
# patch
library(patchwork)
library(cowplot)
theme_set(theme_cowplot(font_size = 10,
                        rel_small = 8/10,
                        rel_tiny = 6/10,
                        rel_large = 10/10,
                        font_family = "Arial"))
# heatmap
library(pheatmap)
library(ComplexHeatmap)
# fonts
library(extrafont)
#font_import()
#font_import(paths = "/T02Data/niuyw/.local/share/fonts")
loadfonts(device = "pdf", quiet = TRUE)
# color
library(ggsci)
# BiocNeighbors
library(BiocNeighbors)
# Seurat
library(Seurat)

# monocle
library(monocle)
```

```{r source, cache = FALSE}
source(here::here("code/preprocess.R"))
source(here::here("code/color.R"))
source(here::here("code/plot.R"))
```

## Load data

Load cell meta

```{r load-cell-meta, cache=FALSE, message=FALSE, warning=FALSE}
cell_meta = read_csv(here::here("output/04.rm_cells/seurat_epithelia.cellmeta.csv")) %>%
  dplyr::select(barcode, seurat_clusters)
head(cell_meta)
```

Load TF activity

```{r}
aucell_res_path = '/niuyw-usb-disk/Projects/scESCA/pySCENIC/200227_6samples'

# SCENIC TFs
aucell_tf = read_csv(file.path(aucell_res_path, 'seurat_epithelia.TF.auc_mtx.csv'))
aucell_tf = aucell_tf %>% 
  mutate(Regulon = str_replace(Regulon, "[(][+][)]", "")) %>% 
  filter(Regulon %in% c("RORA", "RORC", "BARX2", "SOX4")) %>% 
  column_to_rownames(var = 'Regulon')

# t
aucell_tf = as.data.frame(t(aucell_tf)) %>%
  rownames_to_column("barcode")
```

## Random 1 {.tabset}

```{r prepare-cds-random1, eval=FALSE}
seurat = readRDS(here::here('output/04.rm_cells/seurat_epithelia.rds'))
seurat

set.seed(1)
random1 = subset(seurat, downsample=300, seed = NULL)

# Extract data, phenotype data, and feature data from the SeuratObject
expr_matrix = GetAssayData(random1, assay = 'RNA', slot = "counts")
cell_metadata = new("AnnotatedDataFrame", data = random1@meta.data)
gene_annotation = data.frame(gene_short_name = row.names(expr_matrix), row.names = row.names(expr_matrix))
gene_annotation <- new("AnnotatedDataFrame", data = gene_annotation)

# Construct monocle CellDataSet object
cds <- newCellDataSet(expr_matrix, phenoData = cell_metadata, featureData = gene_annotation)

# Normalize the count data
cds <- estimateSizeFactors(cds)

# Calculate dispersions to filter for highly variable genes
cds <- estimateDispersions(cds)

# Constructing Single Cell Trajectories
random1 = FindVariableFeatures(random1, selection.method = 'vst', nfeatures = 2000)
hvg_mask = VariableFeatures(random1)

cds <- setOrderingFilter(cds, hvg_mask)
cds <- reduceDimension(cds, max_components = 2, method = "DDRTree")
cds <- orderCells(cds)

# reverse
cds <- orderCells(cds, reverse = TRUE)

# get DDRtree component
dt = t(reducedDimS(cds))
colnames(dt) = c("DT1", "DT2")

# add dt to seurat obj
random1@reductions$dt = CreateDimReducObject(embeddings = dt, key = "DT_", assay = 'RNA')

# add pseudotime to srat
random1$Pseudotime = pData(cds)$Pseudotime

# save seurat random obj
saveRDS(random1, file = here::here('output/04.rm_cells/seurat_epithelia.random1.rds'))

# and h5ad
sceasy::convertFormat(random1, from="seurat", to="anndata",
                      main_layer = 'counts',
                      outFile=here::here('output/04.rm_cells', paste0('seurat_epithelia.random1', '.h5ad')))

# save for fast reading later
saveRDS(cds, file = here::here('output/04.rm_cells/monocle2_epithelia.random1.rds'))
```

Load cds directly

```{r load-rds-random1, cache=FALSE}
cds1 = readRDS(here::here('output/04.rm_cells/monocle2_epithelia.random1.rds'))
seurat = readRDS(here::here('output/04.rm_cells/seurat_epithelia.random1.rds'))

cds1
```

### State/Pseudotime

```{r plot-state-random1, fig.height=7, fig.width=8}
plot_cell_trajectory(cds1, color_by = "State")
```

```{r plot-pseudotime-random1, fig.height=7, fig.width=8}
plot_cell_trajectory(cds1, color_by = "Pseudotime")
```

### Cluster {.tabset}

```{r plot-cluster-random1, fig.height=8, fig.width=8}
dt = t(reducedDimS(cds1))
colnames(dt) = c("DT1", "DT2")

# Adding cluster label to center of cluster on UMAP
dt_label = dt %>%
  as.data.frame() %>%
  rownames_to_column("barcode") %>%
  left_join(cell_meta, by = "barcode") %>%
  group_by(seurat_clusters) %>%
  summarise(x=median(DT1), y=median(DT2))

plot_cell_trajectory(cds1, color_by = "seurat_clusters", cell_size = 1) +
  geom_text_repel(data = dt_label, aes(label = seurat_clusters, x, y)) +
  scale_color_manual(values = cluster_color_maps) +
  theme(legend.position = "none")
```

### Origin4

```{r plot-Origin4-random1, fig.height=8, fig.width=8}
plot_cell_trajectory(cds1, color_by = "Origin4", cell_size = 2) +
    scale_color_manual(values = origin4_color_maps)
```

### Origin2_n

```{r plot-Origin2_n-random1, fig.height=8, fig.width=8}
plot_cell_trajectory(cds1, color_by = "Origin2_n", cell_size = 2) +
    scale_color_manual(values = origin2_color_maps)
```

### Tissue

```{r plot-tissue-random1, fig.height=8, fig.width=8}
plot_cell_trajectory(cds1, color_by = "Tissue", cell_size = 2) +
    scale_color_manual(values = tissue_color_maps)
```

### Patient

```{r plot-patient-random1, fig.height=8, fig.width=8}
plot_cell_trajectory(cds1, color_by = "Patient", cell_size = 2) +
    scale_color_manual(values = patient_color_maps)
```

### Malignant

```{r plot-Malignant_1-random1, fig.height=8, fig.width=8}
plot_cell_trajectory(cds1, color_by = "malignant", cell_size = 2) +
  scale_color_jco()
```

### Genes

Gene expression

```{r gene-exp-random1, fig.height=9, fig.width=10}
features = c("RORA", "RORC", "BARX2", "SOX4")

FeaturePlot(seurat, features = features, reduction = "dt")
```

TF activity

```{r tf-exp-random1, fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2) & scale_color_gradient(low = "blue", high = "red")
```

### Genes in several clusters

#### 16

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 16), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 13

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 13), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 26

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 26), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 27

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 27), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 29

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 29), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 30

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 30), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

## Random 2 {.tabset}

```{r prepare-cds-random2, eval=FALSE}
seurat = readRDS(here::here('output/04.rm_cells/seurat_epithelia.rds'))
seurat

set.seed(2)
random2 = subset(seurat, downsample=300, seed = NULL)

# Extract data, phenotype data, and feature data from the SeuratObject
expr_matrix = GetAssayData(random2, assay = 'RNA', slot = "counts")
cell_metadata = new("AnnotatedDataFrame", data = random2@meta.data)
gene_annotation = data.frame(gene_short_name = row.names(expr_matrix), row.names = row.names(expr_matrix))
gene_annotation <- new("AnnotatedDataFrame", data = gene_annotation)

# Construct monocle CellDataSet object
cds <- newCellDataSet(expr_matrix, phenoData = cell_metadata, featureData = gene_annotation)

# Normalize the count data
cds <- estimateSizeFactors(cds)

# Calculate dispersions to filter for highly variable genes
cds <- estimateDispersions(cds)

# Constructing Single Cell Trajectories
random2 = FindVariableFeatures(random2, selection.method = 'vst', nfeatures = 2000)
hvg_mask = VariableFeatures(random2)

cds <- setOrderingFilter(cds, hvg_mask)
cds <- reduceDimension(cds, max_components = 2, method = "DDRTree")
cds <- orderCells(cds)

# reverse
cds <- orderCells(cds, reverse = TRUE)

# get DDRtree component
dt = t(reducedDimS(cds))
colnames(dt) = c("DT1", "DT2")

# add dt to seurat obj
random2@reductions$dt = CreateDimReducObject(embeddings = dt, key = "DT_", assay = 'RNA')

# add pseudotime to srat
random2$Pseudotime = pData(cds)$Pseudotime

# save seurat random obj
saveRDS(random2, file = here::here('output/04.rm_cells/seurat_epithelia.random2.rds'))

# and h5ad
sceasy::convertFormat(random2, from="seurat", to="anndata",
                      main_layer = 'counts',
                      outFile=here::here('output/04.rm_cells', paste0('seurat_epithelia.random2', '.h5ad')))

# save for fast reading later
saveRDS(cds, file = here::here('output/04.rm_cells/monocle2_epithelia.random2.rds'))
```

Load cds directly

```{r load-rds-random2, cache=FALSE}
cds2 = readRDS(here::here('output/04.rm_cells/monocle2_epithelia.random2.rds'))
seurat = readRDS(here::here('output/04.rm_cells/seurat_epithelia.random2.rds'))

cds2
```

### State/Pseudotime

```{r plot-state-random2, fig.height=7, fig.width=8}
plot_cell_trajectory(cds2, color_by = "State")
```

```{r plot-pseudotime-random2, fig.height=7, fig.width=8}
plot_cell_trajectory(cds2, color_by = "Pseudotime")
```

### Cluster {.tabset}

```{r plot-cluster-random2, fig.height=8, fig.width=8}
dt = t(reducedDimS(cds2))
colnames(dt) = c("DT1", "DT2")

# Adding cluster label to center of cluster on UMAP
dt_label = dt %>%
  as.data.frame() %>%
  rownames_to_column("barcode") %>%
  left_join(cell_meta, by = "barcode") %>%
  group_by(seurat_clusters) %>%
  summarise(x=median(DT1), y=median(DT2))

plot_cell_trajectory(cds2, color_by = "seurat_clusters", cell_size = 1) +
  geom_text_repel(data = dt_label, aes(label = seurat_clusters, x, y)) +
  scale_color_manual(values = cluster_color_maps) +
  theme(legend.position = "none")
```

### Origin4

```{r plot-Origin4-random2, fig.height=8, fig.width=8}
plot_cell_trajectory(cds2, color_by = "Origin4", cell_size = 2) +
    scale_color_manual(values = origin4_color_maps)
```

### Origin2_n

```{r plot-Origin2_n-random2, fig.height=8, fig.width=8}
plot_cell_trajectory(cds2, color_by = "Origin2_n", cell_size = 2) +
    scale_color_manual(values = origin2_color_maps)
```

### Tissue

```{r plot-tissue-random2, fig.height=8, fig.width=8}
plot_cell_trajectory(cds2, color_by = "Tissue", cell_size = 2) +
    scale_color_manual(values = tissue_color_maps)
```

### Patient

```{r plot-patient-random2, fig.height=8, fig.width=8}
plot_cell_trajectory(cds2, color_by = "Patient", cell_size = 2) +
    scale_color_manual(values = patient_color_maps)
```

### Malignant

```{r plot-Malignant_1-random2, fig.height=8, fig.width=8}
plot_cell_trajectory(cds2, color_by = "malignant", cell_size = 2) +
  scale_color_jco()
```

### Genes

Gene expression

```{r gene-exp-random2, fig.height=9, fig.width=10}
features = c("RORA", "RORC", "BARX2", "SOX4")

FeaturePlot(seurat, features = features, reduction = "dt")
```

TF activity

```{r tf-exp-random2, fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2) & scale_color_gradient(low = "blue", high = "red")
```

### Genes in several clusters

#### 16

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 16), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 13

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 13), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 26

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 26), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 27

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 27), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 29

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 29), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 30

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 30), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

## Random 3 {.tabset}

```{r prepare-cds-random3, eval=FALSE}
seurat = readRDS(here::here('output/04.rm_cells/seurat_epithelia.rds'))
seurat

set.seed(3)
random3 = subset(seurat, downsample=300, seed = NULL)

# Extract data, phenotype data, and feature data from the SeuratObject
expr_matrix = GetAssayData(random3, assay = 'RNA', slot = "counts")
cell_metadata = new("AnnotatedDataFrame", data = random3@meta.data)
gene_annotation = data.frame(gene_short_name = row.names(expr_matrix), row.names = row.names(expr_matrix))
gene_annotation <- new("AnnotatedDataFrame", data = gene_annotation)

# Construct monocle CellDataSet object
cds <- newCellDataSet(expr_matrix, phenoData = cell_metadata, featureData = gene_annotation)

# Normalize the count data
cds <- estimateSizeFactors(cds)

# Calculate dispersions to filter for highly variable genes
cds <- estimateDispersions(cds)

# Constructing Single Cell Trajectories
random3 = FindVariableFeatures(random3, selection.method = 'vst', nfeatures = 2000)
hvg_mask = VariableFeatures(random3)

cds <- setOrderingFilter(cds, hvg_mask)
cds <- reduceDimension(cds, max_components = 2, method = "DDRTree")
cds <- orderCells(cds)

# get DDRtree component
dt = t(reducedDimS(cds))
colnames(dt) = c("DT1", "DT2")

# add dt to seurat obj
random3@reductions$dt = CreateDimReducObject(embeddings = dt, key = "DT_", assay = 'RNA')

# add pseudotime to srat
random3$Pseudotime = pData(cds)$Pseudotime

# save seurat random obj
saveRDS(random3, file = here::here('output/04.rm_cells/seurat_epithelia.random3.rds'))

# and h5ad
sceasy::convertFormat(random3, from="seurat", to="anndata",
                      main_layer = 'counts',
                      outFile=here::here('output/04.rm_cells', paste0('seurat_epithelia.random3', '.h5ad')))

# save for fast reading later
saveRDS(cds, file = here::here('output/04.rm_cells/monocle2_epithelia.random3.rds'))
```

Load cds directly

```{r load-rds-random3, cache=FALSE}
cds3 = readRDS(here::here('output/04.rm_cells/monocle2_epithelia.random3.rds'))
seurat = readRDS(here::here('output/04.rm_cells/seurat_epithelia.random3.rds'))

cds3
```

### State/Pseudotime

```{r plot-state-random3, fig.height=7, fig.width=8}
plot_cell_trajectory(cds3, color_by = "State")
```

```{r plot-pseudotime-random3, fig.height=7, fig.width=8}
plot_cell_trajectory(cds3, color_by = "Pseudotime")
```

### Cluster {.tabset}

```{r plot-cluster-random3, fig.height=8, fig.width=8}
dt = t(reducedDimS(cds3))
colnames(dt) = c("DT1", "DT2")

# Adding cluster label to center of cluster on UMAP
dt_label = dt %>%
  as.data.frame() %>%
  rownames_to_column("barcode") %>%
  left_join(cell_meta, by = "barcode") %>%
  group_by(seurat_clusters) %>%
  summarise(x=median(DT1), y=median(DT2))

plot_cell_trajectory(cds3, color_by = "seurat_clusters", cell_size = 1) +
  geom_text_repel(data = dt_label, aes(label = seurat_clusters, x, y)) +
  scale_color_manual(values = cluster_color_maps) +
  theme(legend.position = "none")
```

### Origin4

```{r plot-Origin4-random3, fig.height=8, fig.width=8}
plot_cell_trajectory(cds3, color_by = "Origin4", cell_size = 2) +
    scale_color_manual(values = origin4_color_maps)
```

### Origin2_n

```{r plot-Origin2_n-random3, fig.height=8, fig.width=8}
plot_cell_trajectory(cds3, color_by = "Origin2_n", cell_size = 2) +
    scale_color_manual(values = origin2_color_maps)
```

### Tissue

```{r plot-tissue-random3, fig.height=8, fig.width=8}
plot_cell_trajectory(cds3, color_by = "Tissue", cell_size = 2) +
    scale_color_manual(values = tissue_color_maps)
```

### Patient

```{r plot-patient-random3, fig.height=8, fig.width=8}
plot_cell_trajectory(cds3, color_by = "Patient", cell_size = 2) +
    scale_color_manual(values = patient_color_maps)
```

### Malignant_1

```{r plot-Malignant_1-random3, fig.height=8, fig.width=8}
plot_cell_trajectory(cds3, color_by = "malignant", cell_size = 2) +
  scale_color_jco()
```

### Genes

Gene expression

```{r gene-exp-random3, fig.height=9, fig.width=10}
features = c("RORA", "RORC", "BARX2", "SOX4")

FeaturePlot(seurat, features = features, reduction = "dt")
```

TF activity

```{r tf-exp-random3, fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2) & scale_color_gradient(low = "blue", high = "red")
```

### Genes in several clusters

#### 16

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 16), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 16) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 13

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 13), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 13) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 26

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 26), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 26) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 27

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 27), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 27) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 29

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 29), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 29) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```

#### 30

Gene expression

```{r , fig.height=9, fig.width=10}
FeaturePlot(subset(seurat, subset = seurat_clusters == 30), features = features, reduction = "dt")
```

TF activity

```{r , fig.height=8, fig.width=10}
dat = FetchData(seurat, vars = c("DT_1", "DT_2", "seurat_clusters")) %>%
  rownames_to_column("barcode") %>%
  left_join(aucell_tf, by = "barcode")

p1 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORA)) +
  geom_point(size = .5)
p2 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = RORC)) +
  geom_point(size = .5)
p3 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = BARX2)) +
  geom_point(size = .5)
p4 = dat %>%
  filter(seurat_clusters == 30) %>%
  ggplot(aes(x = DT_1, y = DT_2, color = SOX4)) +
  geom_point(size = .5)
p1 + p2 + p3 + p4 +
  plot_layout(ncol = 2)
```


## Session info



